pkg/watermark/fetch/interface.go

```diff
@@ -26,8 +26,11 @@ import (
 // Fetcher fetches watermark data from Vn-1 vertex.
 type Fetcher interface {
 	io.Closer
-	// GetWatermark returns the inorder monotonically increasing watermark of the edge connected to Vn-1.
-	GetWatermark(offset isb.Offset, fromPartitionIdx int32) wmb.Watermark
+	// GetWatermark returns the current watermark based on what has been processed
+	GetWatermark() wmb.Watermark
+	// UpdateAndFetchWatermark processes the offset on the partition indicated and returns the overall Watermark
+	// from all Partitions
+	UpdateAndFetchWatermark(offset isb.Offset, fromPartitionIdx int32) wmb.Watermark
```

Comment:
For `UpdateAndFetchWatermark` it seems we use one interface to do two things, both writing and reading. For writing, we are using the input parameters but for reading, it doesn't seem care about the input. Interface users could by mistake think the method is fetching from a certain offset/partitionIdx instead of overall watermark.

Would it be better if we separate them? To have `UpdateWatermark(offset, fromPartitionIdx)` and use the existing `GetWatermark()` for fetching.

Reply:
_Interface users could by mistake think the method is fetching from a certain offset/partitionIdx instead of overall watermark_

That was in fact my motivation for replacing the original `GetWatermark(offset, fromPartitionIdx)`.

I also wanted to have two separate methods like you're saying but @jy4096 expressed a desire to still have a single method to do both like we did before since they're generally done together.

I'll make it whatever the team wants in the end.

Reply:
cc @vigith

Reply:
How about we update 
```
	// GetWatermark returns the current watermark based on what has been processed
	GetWatermark() wmb.Watermark
	// UpdateAndFetchWatermark processes the offset on the partition indicated and returns the overall Watermark
	// from all Partitions
	UpdateAndFetchWatermark(offset isb.Offset, fromPartitionIdx int32) wmb.Watermark
```
to
```
	// GetSmallestLastProcessedWatermark returns the overall smallest last processed watermark
	GetSmallestLastProcessedWatermark() wmb.Watermark
	// ComputeWatermark uses the given offset and partition to compute and return the current watermark
	ComputeWatermark(offset isb.Offset, fromPartitionIdx int32) wmb.Watermark
```
And then in edgeFetcher write a godoc like this
```
        // ComputeWatermark uses the given offset and partition to calculate the watermark, 
        // update the lastProcessedWatermark of the given partition to the watermark we just calculate, 
        // and return the minimum lastProcessedWatermark from all partitions as the watermark. 
        // Finally, it deletes the processor if it's not active.
```

Reply:
The original `GetWatermark(offset)` before we introduced multi partition is also more like "compute and return the watermark for the given offset".

Reply:
Before multi partition, we had a method `GetWatermark(offset) watermark`. This function used the offset from `Read` to calculate and return the watermark for the current readBatch.
With the introduction of multi partition, we updated the method to be `GetWatermark(offset, partition) watermark`. This method had two steps: calculate the watermark for the given offset and partition provided by `readABatch`; compare the last processed watermark among all the partitions and return the smallest value.

In this PR, at first, we extracted the above two steps into two functions: `ProcessOffset(offset, partition) watermark` and `GetWatermark()`.

I think that, when we read a batch and need to assign a watermark to this batch, the operation `ProcessOffset` and `GetWatermark` are strictly bonded together. Therefore, we need to have a new function `Something` (we have been discussing the function name for a long time so I will just use `Something` here).
```
func Something(offset, partition) watermark {
   ProcessOffset(offset, partition) watermark
   GetWatermark()
}
```
The `Read` (which imports the watermark) doesn't need to know that it has to `ProcessOffset to update the last processed watermark` and only then it can `Get the correct watermark for the batch`.

Above is why I prefer to have this `Something` function.

Reply:
@jy4096 thanks for catching me up on the history. +1 on adding more comments which provides ALL important informations about the interface and reduces developer cognitive loads.

I agree that having the `Something` function will simplify the `BatchRead` in a way that we don't need to call two methods in sequence. However, from the view of `Read` (which imports the watermark), I think it needs to know that it has to `UpdateOffset(offset, partition)` first and then `GetWatermark()`, which is exactly why we name `Something` `UpdateAndFetchWatermark` at first place, to explicitly describe what the method does. In general, an interface should provide ALL important informations that the caller needs to know, otherwise we introduce unknowns.

Personally I still prefer separating `Update` and `Get` as it's clearer for the interface. But I am also ok with combining if 1. They are called in sequence very frequently and 2. We put in good descriptions for the `Something` interface, as you did above.

Reply:
It looks right to me to have 2 functions:

1. `GetWatermark()` - the watermark displayed in the UI should be this, right?
2. `Something(offset, partition) watermark`, which includes 2 steps (`ProcessOffset` and `GetWatermark`).

Isn't this true?
->  _xxxx think the method is fetching from a certain offset/partitionIdx instead of overall watermark_

Reply:
another data point to help us make the decision is how often we will call `Update` without following by a `Get`.

Reply:
> Isn't this true?
-> xxxx think the method is fetching from a certain offset/partitionIdx instead of overall watermark

By adding interface descriptions we can avoid such confusion.

Reply:
@whynowy this is not true: "the method is fetching from a certain offset/partitionIdx instead of overall watermark". It is fetching overall watermark. Therefore, do you want to change your mind?

Reply:
> another data point to help us make the decision is how often we will call `Update` without following by a `Get`.

we don't have such use case ATM

Reply:
I would still prefer to have `UpdateWatermark(offset, fromPartitionIdx)` and `GetWatermark()` but can be flexible on that.

But if we keep as is, I'm not sure what's unclear about the current comment "UpdateAndFetchWatermark processes the offset on the partition indicated and returns the **overall Watermark from all Partitions**"

Reply:
> But if we keep as is, I'm not sure what's unclear about the current comment "UpdateAndFetchWatermark processes the offset on the partition indicated and returns the **overall Watermark from all Partitions**"

https://github.com/numaproj/numaflow/pull/875#discussion_r1274179798
How about this proposal?

Reply:
+1

Reply:
> @whynowy this is not true: "the method is fetching from a certain offset/partitionIdx instead of overall watermark". It is fetching overall watermark. Therefore, do you want to change your mind?

@juliev0 - could you help me understand this?

Reply:
@whynowy As you can see in the implementation of `EdgeFetcher` for example, [this](https://github.com/juliev0/numaflow/blob/join-vertex-new/pkg/watermark/fetch/edge_fetcher.go#L81) method which has been renamed to `ComputeWatermark(offset, fromPartitionIdx)` gets the smallest watermark from all partitions. This both sets the current watermark for this partition and then gets the overall watermark from all partitions. I thought dividing it into `UpdateWatermark(offset, fromPartitionIdx)` and `GetWatermark()` might make sense, but I think Juanlu would like to expose just a single method in the interface. Vigith, Juanlu, Siddhant, Keran, and I met and decided in the end that with the interface method being sufficiently commented, that satisfies everyone enough.